solr/core/src/java/org/apache/solr/handler/component/SearchComponent.java

```diff
@@ -78,6 +78,17 @@ public void modifyRequest(ResponseBuilder rb, SearchComponent who, ShardRequest
   public void handleResponses(ResponseBuilder rb, ShardRequest sreq) {
   }
 
+  /** Sets the state of the component. Use {@link #getState(ResponseBuilder, Map)} to build a map compatible with this setter. */
+  public void setState(ResponseBuilder rb, Map<String, Map<String,Object>> state) {
+  }
+
+  /**
+   * Gets a snapshot of the state (placing the state in the map using a component specific key) that can be used to set
+   * the state at a later time. The added state should be javabin compatible.
+   */
+  public void getState(ResponseBuilder rb, Map<String, Map<String, Object>> state) {
```

Comment:
See my comment below. I wonder if this should just be an encode/decode that operates on opaque byte arrays instead?

Reply:
I was more concerned with compatibility if/when we ever needed to change these. It seemed like javabin stuff may be a little more annoying there to get right. But I plan on taking a look this week at using that for serialization instead!

Reply:
The Solr `JavaBinCodec` does know how to encode and decode certain special types (like `SolrDocument`, `SolrDocumentList`, `NamedList`, etc.) which we are already getting for free at the top level as it traverses our maps and encodes the underlying values.

The main added cost with my approach is encoding the maps (with their short keys). The values will be encoded the exact same as if I did manually or let the map values be encoded.

The annoying part about using `JavaBinCodec` is on the write side managing ordering and read/write consistency. It's going to stink if we need to modify this stuff with live running requests. We'd need some form of versioning if we wanted to write raw fields one after another. On the read side, we're still going to want a top level map since there is no guarantee of search component order and existence. I'm wondering if it is actually worth the effort...

Reply:
@donjaime Pushed the latest with javabin encoding as much as possible. Still a little worried about versioning but the rest is fine. I did see maybe a 20% reduction in iterative state size (presumably due to per doc overhead which isn't typically present in most of our facet requests - although they'd have their own overhead in repeated terms and other things). LMK what you think.